In [ ]:
%pip install ipython-sql
%pip install sqlalchemy
%pip install pandas
%pip install mysql-connector-python
%pip install pymysql
%pip install cryptography

In [2]:
%reload_ext sql

from sqlalchemy import create_engine

connection_string = "mysql+pymysql://akshayvaithi:Akshay123@localhost:3306/av763proj"

%sql {connection_string}

In [ ]:
query = """

"""

%sql {query} CREATE TABLE Customers ( \
    customerId INT NOT NULL, \
    firstName VARCHAR(50) NOT NULL, \
    lastName VARCHAR(50) NOT NULL, \
    username VARCHAR(50) NOT NULL, \
    password VARCHAR(50) NOT NULL, \
    emailAddress VARCHAR(320) NULL, \
    phoneNumber VARCHAR(10) NULL)

Creates the customers table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE Orders ( \
    orderId INT NOT NULL, \
    customerId INT NOT NULL, \
    orderDate DATE NOT NULL, \
    totalPrice DECIMAL(50,2) NOT NULL)

Creates the orders table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE Payments ( \
    paymentId INT NOT NULL, \
    customerId INT NOT NULL, \
    paymentMethodId INT NOT NULL, \
    creditCardNumber VARCHAR(16) NOT NULL, \
    nameOnCard VARCHAR(100) NOT NULL, \
    expirationDate VARCHAR(7) NOT NULL)

Creates the payments table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE Products ( \
    productId INT NOT NULL, \
    productName VARCHAR(50) NOT NULL, \
    productUnitPrice DECIMAL(50,2) NOT NULL, \
    productDescription VARCHAR(1000) NULL, \
    productCategory VARCHAR(50) NULL)

Creates the products table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE Shipments ( \
    shipmentId INT NOT NULL, \
    orderId INT NOT NULL, \
    street VARCHAR(100) NOT NULL, \
    city VARCHAR(100) NOT NULL, \
    stateId INT NOT NULL, \
    zipCode INT NOT NULL, \
    shippingServiceId INT NOT NULL, \
    shippingMethodId INT NOT NULL, \
    shipmentDate DATE NULL, \
    deliveryDate DATE NULL)

Creates the shipments table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE itemsInOrder ( \
    orderId INT NOT NULL, \
    orderItem INT NOT NULL, \
    productId INT NOT NULL, \
    itemPrice DECIMAL(50,2) NOT NULL, \
    quantity INT NOT NULL)

Creates the itemsInOrder table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE PaymentMethods ( \
    paymentMethodId INT NOT NULL, \
    paymentMethod VARCHAR(100) NOT NULL)

Creates the PaymentMethods table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE States ( \
    stateId INT NOT NULL, \
    state VARCHAR(2) NOT NULL)

Create the States table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE ShippingServices ( \
    shippingServiceId INT NOT NULL, \
    shippingService VARCHAR(100) NOT NULL)

Creates the ShippingServices table.

In [ ]:
query = """

"""
%sql {query} CREATE TABLE ShippingMethods ( \
    shippingMethodId INT NOT NULL, \
    shippingMethod VARCHAR(100) NOT NULL)

Creates the ShippingMethods table.

In [ ]:
query = """

"""
%sql {query} ALTER TABLE Customers ADD PRIMARY KEY (customerId)
%sql {query} ALTER TABLE Orders ADD PRIMARY KEY (orderId)
%sql {query}ALTER TABLE Payments ADD PRIMARY KEY (paymentId)
%sql {query}ALTER TABLE Products ADD PRIMARY KEY (productId)
%sql {query}ALTER TABLE Shipments ADD PRIMARY KEY (shipmentId)
%sql {query}ALTER TABLE itemsInOrder ADD PRIMARY KEY (orderId, orderItem)
%sql {query}ALTER TABLE PaymentMethods ADD PRIMARY KEY (paymentMethodId)
%sql {query}ALTER TABLE States ADD PRIMARY KEY (stateId)
%sql {query}ALTER TABLE ShippingServices ADD PRIMARY KEY (shippingServiceId)
%sql {query}ALTER TABLE ShippingMethods ADD PRIMARY KEY (shippingMethodId)

Defines the primary keys for all of the tables.

In [ ]:
query = """

"""

%sql {query} ALTER TABLE Orders ADD CONSTRAINT FK_Orders_Customers FOREIGN KEY (customerId) REFERENCES Customers (customerId)
%sql {query} ALTER TABLE Payments ADD CONSTRAINT FK_Payments_Customers FOREIGN KEY (customerId) REFERENCES Customers (customerId)
%sql {query} ALTER TABLE Payments ADD CONSTRAINT FK_Payments_PaymentMethods FOREIGN KEY (paymentMethodId) REFERENCES PaymentMethods (paymentMethodId)
%sql {query} ALTER TABLE Shipments ADD CONSTRAINT FK_Shipments_Orders FOREIGN KEY (orderId) REFERENCES Orders (orderId)
%sql {query} ALTER TABLE Shipments ADD CONSTRAINT FK_Shipments_States FOREIGN KEY (stateId) REFERENCES States (stateId)
%sql {query} ALTER TABLE Shipments ADD CONSTRAINT FK_Shipments_ShippingServices FOREIGN KEY (shippingServiceId) REFERENCES ShippingServices (shippingServiceId)
%sql {query} ALTER TABLE Shipments ADD CONSTRAINT FK_Shipments_ShippingMethods FOREIGN KEY (shippingMethodId) REFERENCES ShippingMethods (shippingMethodId)
%sql {query} ALTER TABLE itemsInOrder ADD CONSTRAINT FK_itemsInOrder_Products FOREIGN KEY (productId) REFERENCES Products (productId)
%sql {query} ALTER TABLE itemsInOrder ADD CONSTRAINT FK_itemsInOrder_Orders FOREIGN KEY (orderId) REFERENCES Orders (orderId)


Defines the foreign keys for all of the tables.

In [ ]:
query = """

"""
%sql {query} INSERT INTO Customers(customerId, firstName, lastName, username, password, emailAddress, phoneNumber) VALUES ( 1001, 'Emma', 'Thompson', 'emmathom', 'Emma@123',  'emma.thompson@gmail.com', '3018794625')
%sql {query} INSERT INTO Customers (customerId, firstName, lastName, username, password, emailAddress, phoneNumber) VALUES ( 1002, 'Alexander', 'Lee', 'alexlee', 'Alex@123',  'alexander.lee@gmail.com', '3015432198')
%sql {query} INSERT INTO Customers (customerId, firstName, lastName, username, password, emailAddress, phoneNumber) VALUES ( 1003, 'Sophia', 'Patel', 'sophiapat', 'Sophia@123',  'sophia.patel@gmail.com', '3018576043')
%sql {query} INSERT INTO Customers (customerId, firstName, lastName, username, password, emailAddress, phoneNumber) VALUES ( 1004, 'Lucas', 'Garcia', 'lucasgar', 'Lucas@123',  'lucas.garcia@gmail.com', '6176248173')
%sql {query} INSERT INTO Customers (customerId, firstName, lastName, username, password, emailAddress, phoneNumber) VALUES ( 1005, 'Olivia', 'Williams', 'oliviawill', 'Olivia@123',  'olivia.williams@gmail.com', '6173859267')


Inserts sample data for the Customers table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO Orders (orderId, customerId, orderDate, totalPrice) VALUES ( 2001, 1001, '2023-06-15', 19.99)
%sql {query} INSERT INTO Orders (orderId, customerId, orderDate, totalPrice) VALUES ( 2002, 1002, '2023-08-17', 79.97)
%sql {query} INSERT INTO Orders (orderId, customerId, orderDate, totalPrice) VALUES ( 2003, 1003, '2024-06-19', 199.94)
%sql {query} INSERT INTO Orders (orderId, customerId, orderDate, totalPrice) VALUES ( 2004, 1004, '2024-06-23', 399.9)
%sql {query} INSERT INTO Orders (orderId, customerId, orderDate, totalPrice) VALUES ( 2005, 1005, '2024-08-21', 699.85)


Inserts sample data for the Orders table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3001, 'The Housemaid is Watching', 19.99, 'From New York Times, USA Today and Wall Street Journal bestselling author Freida McFadden comes the next installment of the unbelievably twisty, tension-packed and globally bestselling Housemaid series.', 'Books')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3002, 'A Court of Thorns and Roses', 19.99, 'From bestselling author Sarah J. Maas comes a seductive, breathtaking book that blends romance, adventure, and faerie lore into an unforgettable read.', 'Books')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3003, 'A Game of Thrones 5-Book Boxed Set', 29.99, 'For the first time, all five novels in the epic fantasy series that inspired HBO’s Game of Thrones are together in one eBook bundle.', 'Books')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3004, 'Milk-Bone Original Dog Treats for Medium Dogs', 19.99, 'Wholesome and tasty dog treats - now with even more meaty taste compared to the Milk-Bone Original biscuits you know and love.', 'Pet Supplies')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3005, 'Greenies Original Natural Dental Care Treats', 29.99, 'One Greenies dental treat a day is all it takes for clean teeth, fresh breath, and a happy dog.', 'Pet Supplies')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3006, 'Zesty Paws Dog Allergy Relief', 39.99, 'Made with premium ingredients that support digestion, immune function, and seasonal allergies, these chews are great for your doggie.', 'Pet Supplies')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3007, 'JOGAL Mens Floral Hawaiian Shirt', 19.99, 'Regular fit, short sleeve Button Down Shirt.', 'Clothing, Shoes & Jewelry')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3008, 'Wrangler Men''s Classic Cargo Stretch Short', 29.99, 'Built for versatility and comfort, this classic cargo short takes you from the outdoors to work, to every day with this timeless silhouette.', 'Clothing, Shoes & Jewelry')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3009, 'Haggar Men''s Premium No Iron Khaki', 39.99, 'Stretchable cotton-blend fabric, hidden expandable waistband and permanent crease will keep you looking sharp and comfortable.', 'Clothing, Shoes & Jewelry')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3010, 'GHOST Whey Protein Powder, Chips Ahoy', 49.99, 'Versatile and Delicious: GHOST Whey x Chips Ahoy! combines a premium, fully disclosed whey protein blend, a few digestive enzymes, and real Chips Ahoy! cookie pieces in every scoop for a flavor that is second to none. 26 servings.', 'Grocery & Gourmet Food')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3011, 'Honey Stinger Organic Honey Waffle', 19.99, 'The Honey waffle has a thin layer of honey-infused filling sandwiched between two thin waffles. The perfect combo of gooey filling and crispy waffle, it’s the fuel you’ll crave while on-the-go.', 'Grocery & Gourmet Food')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3012, 'UNREAL Dark Chocolate Coconut Bars', 29.99, 'ONLY 3g OF SUGAR & 3 SIMPLE INGREDIENTS: organic coconut, organic cassava syrup, and dark chocolate.', 'Grocery & Gourmet Food')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3013, 'Uncharted: Legacy of Thieves (Remastered)', 39.99, 'Discover lost history with the charismatic yet complex thieves, Nathan Drake and Chloe Frazer, as they travel the world with a sense of wonder, pursuing extraordinary adventures and lost lore.', 'Video Games')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3014, 'The Last of Us Part II Remastered', 49.99, 'Relive or play for the first time Ellie and Abby''s story, now with graphical enhancements, new gameplay modes like the roguelike survival experience No Return, full DualSense® wireless controller integration, and more.', 'Video Games')
%sql {query} INSERT INTO Products (productId, productName, productUnitPrice, productDescription, productCategory) VALUES ( 3015, 'Marvel''s Spider-Man 2', 59.99, 'Swing, jump and utilize the new Web Wings to travel across Marvel''s New York, quickly switching between Peter Parker and Miles Morales to experience different stories and epic new powers, as the iconic villain Venom threatens to destroy their lives, their city and the ones they love.', 'Video Games')



Inserts sample data for the Products table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2001, 1, 3001, 19.99, 1)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2002, 1, 3002, 19.99, 1)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2002, 2, 3003, 59.98, 2)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2003, 1, 3004, 19.99, 1)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2003, 2, 3005, 59.98, 2)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2003, 3, 3006, 119.97, 3)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2004, 1, 3007, 19.99, 1)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2004, 2, 3008, 59.98, 2)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2004, 3, 3009, 119.97, 3)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2004, 4, 3010, 199.96, 4)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2005, 1, 3011, 19.99, 1)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2005, 2, 3012, 59.98, 2)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2005, 3, 3013, 119.97, 3)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2005, 4, 3014, 199.96, 4)
%sql {query} INSERT INTO itemsInOrder (orderId, orderItem, productId, itemPrice, quantity) VALUES ( 2005, 5, 3015, 299.95, 5)



Inserts sample data for the itemsInOrder table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO PaymentMethods (paymentMethodId, paymentMethod) VALUES ( 6001, 'American Express')
%sql {query} INSERT INTO PaymentMethods (paymentMethodId, paymentMethod) VALUES ( 6002, 'Mastercard')
%sql {query} INSERT INTO PaymentMethods (paymentMethodId, paymentMethod) VALUES ( 6003, 'Discover')
%sql {query} INSERT INTO PaymentMethods (paymentMethodId, paymentMethod) VALUES ( 6004, 'Visa')


Inserts sample data for the PaymentMethods table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO States (stateId, state) VALUES ( 7001, 'ME')
%sql {query} INSERT INTO States (stateId, state) VALUES ( 7002, 'MD')
%sql {query} INSERT INTO States (stateId, state) VALUES ( 7003, 'MA')
%sql {query} INSERT INTO States (stateId, state) VALUES ( 7004, 'MI')
%sql {query} INSERT INTO States (stateId, state) VALUES ( 7005, 'MN')

Inserts sample data for the States table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO ShippingServices (shippingServiceId, shippingService) VALUES ( 8001, 'DHL')
%sql {query} INSERT INTO ShippingServices (shippingServiceId, shippingService) VALUES (  8002, 'USPS')
%sql {query} INSERT INTO ShippingServices (shippingServiceId, shippingService) VALUES (  8003, 'UPS')
%sql {query} INSERT INTO ShippingServices (shippingServiceId, shippingService) VALUES (  8004, 'FedEx')

Inserts sample data for the ShippingServices table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO ShippingMethods (shippingMethodId, shippingMethod) VALUES ( 9001, 'Expedited')
%sql {query} INSERT INTO ShippingMethods (shippingMethodId, shippingMethod) VALUES ( 9002, 'Overnight')
%sql {query} INSERT INTO ShippingMethods (shippingMethodId, shippingMethod) VALUES ( 9003, 'Priority')
%sql {query} INSERT INTO ShippingMethods (shippingMethodId, shippingMethod) VALUES ( 9004, 'Flat-rate')
%sql {query} INSERT INTO ShippingMethods (shippingMethodId, shippingMethod) VALUES ( 9005, 'Economy')

Inserts sample data for the ShippingMethods table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO Payments (paymentId, customerId, paymentMethodId, creditCardNumber, nameOnCard, expirationDate) VALUES ( 4001, 1001, 6002, '2019384756129384', 'Emma Thompson',  '05/2028')
%sql {query} INSERT INTO Payments (paymentId, customerId, paymentMethodId, creditCardNumber, nameOnCard, expirationDate) VALUES ( 4002, 1002, 6004, '4671892304567890', 'Alexander Lee',  '03/2029')
%sql {query} INSERT INTO Payments (paymentId, customerId, paymentMethodId, creditCardNumber, nameOnCard, expirationDate) VALUES ( 4003, 1003, 6002, '2556783491236789', 'Sophia Patel',  '07/2028')
%sql {query} INSERT INTO Payments (paymentId, customerId, paymentMethodId, creditCardNumber, nameOnCard, expirationDate) VALUES ( 4004, 1004, 6004, '4132987654321098', 'Lucas Garcia',  '05/2029')
%sql {query} INSERT INTO Payments (paymentId, customerId, paymentMethodId, creditCardNumber, nameOnCard, expirationDate) VALUES ( 4005, 1005, 6002, '2345678901234567', 'Olivia Williams',  '09/2028')

Inserts sample data for the Payments table.

In [ ]:
query = """

"""
%sql {query} INSERT INTO Shipments (shipmentId, orderId, street, city, stateId, zipCode, shippingServiceId, shippingMethodId, shipmentDate, deliveryDate) VALUES ( 5001, 2001, '39 Oak Street', 'Rockville', 7002,  20850, 8003, 9004, '2023-6-16', '2023-6-18')
%sql {query} INSERT INTO Shipments (shipmentId, orderId, street, city, stateId, zipCode, shippingServiceId, shippingMethodId, shipmentDate, deliveryDate) VALUES ( 5002, 2002, '17 Willow Lane', 'Germantown', 7002,  20874, 8004, 9005, '2023-08-19', '2023-08-21')
%sql {query} INSERT INTO Shipments (shipmentId, orderId, street, city, stateId, zipCode, shippingServiceId, shippingMethodId, shipmentDate, deliveryDate) VALUES ( 5003, 2003, '42 Pine Street', 'Silver Spring', 7002, 20901, 8003, 9004, '2024-06-20', '2024-06-22')
%sql {query} INSERT INTO Shipments (shipmentId, orderId, street, city, stateId, zipCode, shippingServiceId, shippingMethodId, shipmentDate, deliveryDate) VALUES ( 5004, 2004, '78 Elm Avenue', 'Belmont', 7003,  02478, 8004, 9005, '2024-06-25', '2024-06-27')
%sql {query} INSERT INTO Shipments (shipmentId, orderId, street, city, stateId, zipCode, shippingServiceId, shippingMethodId, shipmentDate, deliveryDate) VALUES ( 5005, 2005, '215 Maple Road', 'Chelsea', 7003,  02150, 8003, 9004, '2024-08-22', '2024-08-24')

Inserts sample data for the Shipments table.

In [106]:
query = """

"""
%sql {query} SELECT stateId FROM Shipments GROUP BY stateId

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
2 rows affected.


[(7002,), (7003,)]

How many different states were products shipped to ? : 2

In [107]:
query = """

"""
%sql {query} SELECT productName, productUnitPrice FROM Products ORDER BY productUnitPrice DESC LIMIT 1

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.


[('Marvels Spider-Man 2', Decimal('59.99'))]

What is the highest priced item in the inventory ? : Marvel’s Spider-Man 2

In [108]:
query = """

"""
%sql {query} SELECT firstName,lastName,totalPrice FROM Orders INNER JOIN Customers ON Orders.customerId=Customers.customerId ORDER BY totalPrice DESC LIMIT 1

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.


[('Olivia', 'Williams', Decimal('699.85'))]

Which customer placed the highest priced order ? : Olivia Williams

In [109]:
query = """

"""
%sql {query} SELECT PaymentMethod, COUNT(*) FROM Payments INNER JOIN PaymentMethods ON Payments.paymentMethodId=PaymentMethods.paymentMethodId GROUP BY PaymentMethod ORDER BY COUNT(*) DESC LIMIT 1

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.


[('Mastercard', 3)]

What is the name of the most frequently used payment method by customers ? : Mastercard

In [110]:
query = """

"""
%sql {query} SELECT paymentMethod FROM PaymentMethods LEFT OUTER JOIN Payments ON Payments.paymentMethodId=PaymentMethods.paymentMethodID WHERE paymentID IS NULL

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
2 rows affected.


[('American Express',), ('Discover',)]

What payment methods were not used by any customers ? : American Express, Discover

In [111]:
query = """

"""
%sql {query} SELECT SUM(totalPrice) FROM Orders

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.


[(Decimal('1399.65'),)]

What is the total amount of revenue generated from all the orders ? : $1399.65

In [112]:
query = """

"""
%sql {query} SELECT AVG(totalPrice) FROM Orders

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.


[(Decimal('279.930000'),)]

What is the average cost of an order placed ? : $279.93

In [114]:
query = """

"""
%sql {query} SELECT COUNT(*) FROM Orders

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.


[(5,)]

How many orders were placed ? : 5

In [115]:
query = """

"""
%sql {query} SELECT nameOnCard,paymentMethod FROM Payments INNER JOIN PaymentMethods ON PaymentMethods.paymentMethodId=Payments.paymentMethodId WHERE paymentMethod='Mastercard'

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
3 rows affected.


[('Emma Thompson', 'Mastercard'),
 ('Sophia Patel', 'Mastercard'),
 ('Olivia Williams', 'Mastercard')]

Which customers used Mastercard ? : Emma Thompson, Sophia Patel, Olivia Williams

In [116]:
query = """

"""
%sql {query} SELECT orderId,state FROM Shipments INNER JOIN States ON Shipments.stateId=States.stateId WHERE state='MD'

 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
3 rows affected.


[(2001, 'MD'), (2002, 'MD'), (2003, 'MD')]

Which orders got shipped to Maryland ? : 2001, 2002, 2003

In [117]:
query = """

"""
%sql {query} SELECT orderId,shippingService FROM Shipments INNER JOIN ShippingServices ON Shipments.shippingServiceId=ShippingServices.shippingServiceId WHERE shippingService='UPS'


 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
3 rows affected.


[(2001, 'UPS'), (2003, 'UPS'), (2005, 'UPS')]

Which orders were shipped with UPS ? : 2001,2003,2005

In [118]:
query = """

"""
%sql {query} SELECT shipmentId FROM Shipments WHERE shipmentDate>'2024-01-01'


 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
3 rows affected.


[(5003,), (5004,), (5005,)]

Which shipments were shipped AFTER January 1, 2024 ? : 5003,5004,5005

In [121]:
query = """

"""
%sql {query} UPDATE Shipments SET zipCode = 22478 WHERE orderId= 2004
%sql {query} UPDATE Shipments SET zipCode = 22150 WHERE orderId= 2005
%sql {query} SELECT * FROM Shipments



 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.
 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
1 rows affected.
 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
5 rows affected.


[(5001, 2001, '39 Oak Street', 'Rockville', 7002, 20850, 8003, 9004, datetime.date(2023, 6, 16), datetime.date(2023, 6, 18)),
 (5002, 2002, '17 Willow Lane', 'Germantown', 7002, 20874, 8004, 9005, datetime.date(2023, 8, 19), datetime.date(2023, 8, 21)),
 (5003, 2003, '42 Pine Street', 'Silver Spring', 7002, 20901, 8003, 9004, datetime.date(2024, 6, 20), datetime.date(2024, 6, 22)),
 (5004, 2004, '78 Elm Avenue', 'Belmont', 7003, 22478, 8004, 9005, datetime.date(2024, 6, 25), datetime.date(2024, 6, 27)),
 (5005, 2005, '215 Maple Road', 'Chelsea', 7003, 22150, 8003, 9004, datetime.date(2024, 8, 22), datetime.date(2024, 8, 24))]

How would you update zip codes that were entered incorrectly ?

In [123]:
query = """

"""
%sql {query} UPDATE Shipments SET zipCode = 20878 WHERE stateId = 7002
%sql {query} SELECT * FROM Shipments


 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
3 rows affected.
 * mysql+pymysql://akshayvaithi:***@localhost:3306/av763proj
5 rows affected.


[(5001, 2001, '39 Oak Street', 'Rockville', 7002, 20878, 8003, 9004, datetime.date(2023, 6, 16), datetime.date(2023, 6, 18)),
 (5002, 2002, '17 Willow Lane', 'Germantown', 7002, 20878, 8004, 9005, datetime.date(2023, 8, 19), datetime.date(2023, 8, 21)),
 (5003, 2003, '42 Pine Street', 'Silver Spring', 7002, 20878, 8003, 9004, datetime.date(2024, 6, 20), datetime.date(2024, 6, 22)),
 (5004, 2004, '78 Elm Avenue', 'Belmont', 7003, 22478, 8004, 9005, datetime.date(2024, 6, 25), datetime.date(2024, 6, 27)),
 (5005, 2005, '215 Maple Road', 'Chelsea', 7003, 22150, 8003, 9004, datetime.date(2024, 8, 22), datetime.date(2024, 8, 24))]

How would you update the zip codes of a group of customers from the same region in Maryland? 